File: hw7/GTUSet.java
import java.security.InvalidParameterException;

/**
 * Create a set whose type is T 
 * @author ferhat
 * @param <T> T is type of set
 */
public class GTUSet<T> implements GTUSetInt<T> {
    
    private final T[] data;
    private int Dsize; 
    private final int Dcapacity =100;
    
    /**
     * Constructor of GTUSet 
     */
    public GTUSet(){
        data = (T[])new Object[Dcapacity];
        Dsize =0;
    }
    
    /**
     *
     * @return true if The set is empty otherwise false.
     */
    @Override
    public boolean empty(){
        
        return Dsize == 0;
    }
    
    /**
     *
     * @return size of the Set
     */
    @Override
    public int size(){
   
        return Dsize;
    }
    
    /**
     *
     * @return maximum size that Set can have
     */
    @Override
    public int max_size(){
     
        return Dcapacity;
    }
    
    /**
     *
     * @param element which is type of T added to set 
     * @throws InvalidParameterException If element is already in set then throws exception
     */
    @Override
    public void insert(T element)throws InvalidParameterException{
    
        if(find(element).equals(end())){ // if argument does not exist then insert it
        
            if(Dsize < Dcapacity){
 
                data[Dsize] =element;
                ++Dsize;
            }
            else{
            
                System.err.println("Capacity is full!!!");
            }
        
        }
        else{
            throw new InvalidParameterException();     
        }
    }
    
    /**
     *
     * @param other set to compare with itself 
     * @return Set of intersection of itself and other set
     */
    @Override
    public GTUSetInt<T> intersection(GTUSetInt<T> other){
    
        GTUSetInt<T> temp;
        temp = new GTUSet<>();
        for(int i=0;i<Dsize ;++i){
        
            if(!other.find(data[i]).equals(other.end())){ // looking if element exist in both sets
            
                temp.insert(data[i]);
            }      
        }        
        return temp;
    }

    /**
     *  if element exist in the set then erase it    
     * @param element to be erased
     */
    @Override
    @SuppressWarnings("empty-statement")
    public void erase(T element){
        
        if(!find(element).equals(end())){  // if it is in the set then erase it
        
            int i=0;
            for(;!data[i].equals(element) ;++i); // looking for its index
            
            for(;i<Dsize-1;++i){
            
                data[i] =data[i+1];
            }
            --Dsize;   
        }
    
    }

    /**
     * Delete all elements in the set 
     * set Size to 0
     */
    @Override
    public void clear(){
        Dsize =0;
    }
    
    /**
     *
     * @param element that is looked
     * @return because all elements  in a set are unique   if element is in the set then return 1 otherwise 0 
     */
    @Override
    public int count(T element){
    
        if(!find(element).equals(end())){  
                                            
            return 1;
        }
        else{ 
            return 0;
        }
    }
    
    /**
     *
     * @param element to be found
     * @return Iterator if exists 
     */
    @Override
    public GTUIterator find(T element){
        int i =0;   
        for(;i<Dsize;++i){
        
            if(data[i].equals(element)){
            
                break;
            }
        }
        return new GTUIterator(data,Dsize,i);
    }

    /**
     *
     * @return Iterator for first element of set
     */
    @Override
    public GTUIterator begin(){
        return new GTUIterator(data,Dsize,0);
    }

    /**
     *
     * @return Iterator for end of the set
     */
    @Override
    public GTUIterator end(){
    
        return new GTUIterator(data,Dsize,Dsize);
    }
}

File: hw6/GTUSet.cpp
#include "GTUSet.h"
#include<iostream>
using namespace std;
namespace GTUFerhat{
    
    template<class T>
    GTUSet<T>::GTUSet():GTUSetBase<T>(),data(new T[100],default_delete<T[]>()),Dsize(0),Dcapacity(100){
        
    }
    
    template<class T>
    bool GTUSet<T>::empty()const{
        if(Dsize == 0){
            return true;
        }
        else
            return false;
    }
    
    template<class T>
    int GTUSet<T>::size()const{
        return Dsize;
    }
    
    template<class T>
    int GTUSet<T>::max_size()const{
        return Dcapacity;  // capacity of array of shared_ptr 
    }
    
    template<class T>
    void GTUSet<T>::insert(const T& v)throw (invalid_argument){ // if argument is already in set then throw invalid _argument 
         
        if(find(v) == end()){  // if argument does not exist then insert it
          
            if(Dsize <= Dcapacity ){
                int i;
                for(i=0;i<Dsize;++i){   // to make them in order
                    if(v < data.get()[i]){
                        break;
                    }
                }
                for(int j =Dsize;i < j;--j){
                    data.get()[j] =data.get()[j-1];
                }
                data.get()[i] =v;
                ++Dsize;
            }
            else{
                cerr<<"Capacity is full\n";
            }
        }
        else{
            throw invalid_argument("The element is already in set!!!\n");
        }
    } 
    
    template<class T>
    void GTUSet<T>::erase(const T& v){
        if(find(v) != end()){ // if it is in the set then erase it
            int i;
            for(i=0;data.get()[i] !=v ;++i); // looking for its index
            
            for(;i < Dsize-1 ;++i){
              data.get()[i] = data.get()[i+1];
          }          
          --Dsize;
        }
    }
    
    template<class T>
    void GTUSet<T>::clear(){
        Dsize =0;
    }
    
    template<class T>
    typename GTUSetBase<T>::GTUIterator GTUSet<T>::find(const T& v){ // return type is GTUIterator 
        int i;                                                       
        for(i=0;i<Dsize;++i){
            if(data.get()[i] == v){
                break;
            }
        }
        typename GTUSetBase<T>::GTUIterator temp(&data.get()[i]);  // making temp GTUIterator obj to return
        return temp;
    }
    
    template<class T>
    int GTUSet<T>::count(const T& v){ // because all elements  in a set are unique  
        if(find(v) !=end()){          // if element is in the set then return 1 otherwise 0
            return 1;
        }
        else
            return 0;
    }
    
    template<class T>
    typename GTUSetBase<T>::GTUIterator GTUSet<T>::begin(){ // returning the address of first element type of GTUIterator
        
        typename GTUSetBase<T>::GTUIterator temp(&data.get()[0]);
    
        return temp;
    }
    
    template<class T>
    typename GTUSetBase<T>::GTUIterator GTUSet<T>::end(){ // returning the address of last+1  element type of GTUIterator
        
   
        typename GTUSetBase<T>::GTUIterator v(&data.get()[Dsize]);
        return v;
    }
    
}
